src/lib/FlightTasks/tasks/Utility/VelocitySmoothing.cpp

```diff
@@ -171,112 +154,132 @@ float VelocitySmoothing::computeT2(float T123, float T1, float T3)
+	int direction = math::sign(_vel_sp - vel_zero_acc);
 
-	// compute increasing acceleration time
-	if (T123 < 0.f) {
-		T1 = computeT1(_accel, _vel, _vel_sp, _max_jerk_T1);
-
-	} else {
-		T1 = computeT1(T123, _accel, _vel, _vel_sp, _max_jerk_T1);
+	if (direction == 0) {
+		// If by braking immediately the velocity is exactly
+		// the require one with zero acceleration, then brake
+		direction = math::sign(_state.a);
 	}
 
-	// compute decreasing acceleration time
-	T3 = computeT3(T1, _accel, _max_jerk_T1);
+	return direction;
+}
 
-	// compute constant acceleration time
-	if (T123 < 0.f) {
-		T2 = computeT2(T1, T3, _accel, _vel, _vel_sp, _max_jerk_T1);
+float VelocitySmoothing::computeVelAtZeroAcc()
+{
+	float vel_zero_acc = _state.v;
 
-	} else {
-		T2 = computeT2(T123, T1, T3);
+	if (fabsf(_state.a) > FLT_EPSILON) {
+		float j_zero_acc = -math::sign(_state.a) * _max_jerk; // Required jerk to reduce the acceleration
+		float t_zero_acc = -_state.a / j_zero_acc; // Required time to cancel the current acceleration
+		vel_zero_acc = _state.v + _state.a * t_zero_acc + 0.5f * j_zero_acc * t_zero_acc * t_zero_acc;
 	}
 
-	_T1 = T1;
-	_T2 = T2;
-	_T3 = T3;
+	return vel_zero_acc;
 }
 
-void VelocitySmoothing::integrate(float &accel_setpoint_smooth, float &vel_setpoint_smooth,
-				  float &pos_setpoint_smooth)
+void VelocitySmoothing::updateDurationsMinimizeTotalTime()
 {
-	integrate(_dt, 1.f, accel_setpoint_smooth, vel_setpoint_smooth, pos_setpoint_smooth);
+	float jerk_max_T1 = _direction * _max_jerk;
+	float delta_v = _vel_sp - _state.v;
+
+	// compute increasing acceleration time
+	_T1 = computeT1(_state.a, delta_v, jerk_max_T1, _max_accel);
+
+	// compute decreasing acceleration time
+	_T3 = computeT3(_T1, _state.a, jerk_max_T1);
+
+	// compute constant acceleration time
+	_T2 = computeT2(_T1, _T3, _state.a, delta_v, jerk_max_T1);
 }
 
-void VelocitySmoothing::integrate(float dt, float integration_scale_factor, float &accel_setpoint_smooth,
-				  float &vel_setpoint_smooth,
-				  float &pos_setpoint_smooth)
+Trajectory VelocitySmoothing::evaluatePoly(float j, float a0, float v0, float x0, float t, int d)
 {
-	/* Apply correct jerk (min, max or zero) */
-	if (_T1 > FLT_EPSILON) {
-		_jerk = _max_jerk_T1;
+	Trajectory traj;
+	float jt = d * j;
+	float t2 = t * t;
+	float t3 = t2 * t;
 
-		if (_T1 < dt && dt > _dt) {
-			// _T1 was supposed to be _dt, however, now, dt is bigger than _dt. We have to reduce the jerk to avoid an acceleration overshoot.
-			_jerk *= _dt / dt; // Keep the same area _dt * _jerk = dt * jerk_new
-		}
+	traj.j = jt;
+	traj.a = a0 + jt * t;
+	traj.v = v0 + a0 * t + 0.5f * jt * t2;
+	traj.x = x0 + v0 * t + 0.5f * a0 * t2 + 1.f / 6.f * jt * t3;
 
-	} else if (_T2 > FLT_EPSILON) {
-		_jerk = 0.f;
+	return traj;
+}
 
-	} else if (_T3 > FLT_EPSILON) {
-		_jerk = -_max_jerk_T1;
+void VelocitySmoothing::updateTraj(float dt, float time_stretch)
+{
+	_local_time += dt * time_stretch;
+	const float t = _local_time;
+	float t1 = 0.f;
```

Comment:
With the `t` spiltup I suggested you could initialize these when you know the value and only if they are needed.